File: backend/src/routes/prescriptions.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const pool = require('../db');
const { requireAuth } = require('../middleware/auth');

const router = express.Router();

// GET /prescriptions
router.get('/', requireAuth, async (req, res) => {
  try {
    const user = req.user;
    let result;

    if (user.role === 'patient') {
      result = await pool.query(
        `SELECT * FROM prescriptions WHERE patient_id = $1 ORDER BY prescribed_date DESC`,
        [user.id]
      );
    } else if (user.role === 'doctor') {
      result = await pool.query(
        `SELECT * FROM prescriptions WHERE doctor_id = $1 ORDER BY prescribed_date DESC`,
        [user.id]
      );
    } else {
      return res.status(403).json({ error: 'Unauthorized' });
    }

    const prescriptions = result.rows.map(mapPrescription);
    return res.json({
      prescriptions,
      totalCount: prescriptions.length,
      activeCount: prescriptions.filter(p => p.status === 'active').length,
      completedCount: prescriptions.filter(p => p.status === 'completed').length,
    });
  } catch (err) {
    console.error('Fetch prescriptions error:', err);
    return res.status(500).json({ error: 'Failed to fetch prescriptions' });
  }
});

// GET /prescriptions/:id
router.get('/:id', requireAuth, async (req, res) => {
  try {
    const user = req.user;
    const { id } = req.params;

    const result = await pool.query(
      `SELECT * FROM prescriptions WHERE id = $1`,
      [id]
    );
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Prescription not found' });
    }

    const p = result.rows[0];
    // Only doctor or patient involved can view
    if (p.patient_id !== user.id && p.doctor_id !== user.id) {
      return res.status(403).json({ error: 'Unauthorized' });
    }

    return res.json(mapPrescription(p));
  } catch (err) {
    console.error('Fetch prescription error:', err);
    return res.status(500).json({ error: 'Failed to fetch prescription' });
  }
});

// POST /prescriptions - create prescription (doctor only)
router.post('/', requireAuth, async (req, res) => {
  try {
    const user = req.user;
    if (user.role !== 'doctor') {
      return res.status(403).json({ error: 'Only doctors can create prescriptions' });
    }

    const { patientId, patientName, diagnosis, medicines, labTests, instructions, followUpDate, consultationFee } = req.body;
    if (!patientId) {
      return res.status(400).json({ error: 'patientId is required' });
    }

    const id = uuidv4();
    const result = await pool.query(
      `INSERT INTO prescriptions (
        id, patient_id, doctor_id, patient_name, doctor_name, doctor_specialization,
        diagnosis, medicines, lab_tests, instructions, follow_up_date, consultation_fee, status
      ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,'active')
      RETURNING *`,
      [
        id, patientId, user.id, patientName || null,
        user.full_name, user.specialization || 'General Physician',
        diagnosis || null,
        JSON.stringify(medicines || []),
        JSON.stringify(labTests || []),
        instructions || null,
        followUpDate || null,
        consultationFee || null,
      ]
    );

    return res.status(201).json({
      success: true,
      prescription: mapPrescription(result.rows[0]),
    });
  } catch (err) {
    console.error('Create prescription error:', err);
    return res.status(500).json({ error: 'Failed to create prescription' });
  }
});

function mapPrescription(row) {
  return {
    id: row.id,
    patientId: row.patient_id,
    doctorId: row.doctor_id,
    patientName: row.patient_name,
    doctorName: row.doctor_name,
    doctorSpecialization: row.doctor_specialization,
    diagnosis: row.diagnosis,
    medicines: row.medicines || [],
    labTests: row.lab_tests || [],
    instructions: row.instructions,
    followUpDate: row.follow_up_date,
    consultationFee: row.consultation_fee,
    status: row.status,
    prescribedDate: row.prescribed_date,
    createdAt: row.created_at,
  };
}

module.exports = router;

File: backend/routers/prescriptions.py
"""
routers/prescriptions.py — Prescription CRUD routes for Clynicx Python backend.
Mirrors the behaviour of the original Node.js src/routes/prescriptions.js.

Endpoints:
    GET  /prescriptions
    GET  /prescriptions/{prescription_id}
    POST /prescriptions
"""
import uuid
import json
from typing import Optional, List, Any
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from db import query
from middleware.auth import require_auth

router = APIRouter(prefix="/prescriptions", tags=["prescriptions"])


def _map_prescription(row: dict) -> dict:
    medicines = row.get("medicines") or []
    lab_tests = row.get("lab_tests") or []

    # psycopg2 returns JSONB as a Python object already; handle string case too
    if isinstance(medicines, str):
        try:
            medicines = json.loads(medicines)
        except Exception:
            medicines = []
    if isinstance(lab_tests, str):
        try:
            lab_tests = json.loads(lab_tests)
        except Exception:
            lab_tests = []

    follow_up = row.get("follow_up_date")
    if follow_up is not None:
        follow_up = str(follow_up)

    return {
        "id": str(row["id"]),
        "patientId": str(row["patient_id"]),
        "doctorId": str(row["doctor_id"]),
        "patientName": row.get("patient_name"),
        "doctorName": row.get("doctor_name"),
        "doctorSpecialization": row.get("doctor_specialization"),
        "diagnosis": row.get("diagnosis"),
        "medicines": medicines,
        "labTests": lab_tests,
        "instructions": row.get("instructions"),
        "followUpDate": follow_up,
        "consultationFee": row.get("consultation_fee"),
        "status": row.get("status"),
        "prescribedDate": str(row["prescribed_date"]) if row.get("prescribed_date") else None,
        "createdAt": str(row["created_at"]) if row.get("created_at") else None,
    }


# ── Request Body ──────────────────────────────────────────────────────────────

class CreatePrescriptionBody(BaseModel):
    patientId: str
    patientName: Optional[str] = None
    diagnosis: Optional[str] = None
    medicines: Optional[List[Any]] = []
    labTests: Optional[List[Any]] = []
    instructions: Optional[str] = None
    followUpDate: Optional[str] = None
    consultationFee: Optional[Any] = None


# ── GET /prescriptions ────────────────────────────────────────────────────────

@router.get("")
def list_prescriptions(user: dict = Depends(require_auth)):
    if user["role"] == "patient":
        rows = query(
            "SELECT * FROM prescriptions WHERE patient_id = %s ORDER BY prescribed_date DESC",
            (user["id"],),
        )
    elif user["role"] == "doctor":
        rows = query(
            "SELECT * FROM prescriptions WHERE doctor_id = %s ORDER BY prescribed_date DESC",
            (user["id"],),
        )
    else:
        raise HTTPException(status_code=403, detail="Unauthorized")

    prescriptions = [_map_prescription(r) for r in rows]
    return {
        "prescriptions": prescriptions,
        "totalCount": len(prescriptions),
        "activeCount": sum(1 for p in prescriptions if p["status"] == "active"),
        "completedCount": sum(1 for p in prescriptions if p["status"] == "completed"),
    }


# ── GET /prescriptions/{prescription_id} ─────────────────────────────────────

@router.get("/{prescription_id}")
def get_prescription(prescription_id: str, user: dict = Depends(require_auth)):
    rows = query("SELECT * FROM prescriptions WHERE id = %s", (prescription_id,))
    if not rows:
        raise HTTPException(status_code=404, detail="Prescription not found")

    p = rows[0]
    if str(p["patient_id"]) != str(user["id"]) and str(p["doctor_id"]) != str(user["id"]):
        raise HTTPException(status_code=403, detail="Unauthorized")

    return _map_prescription(p)


# ── POST /prescriptions ───────────────────────────────────────────────────────

@router.post("", status_code=201)
def create_prescription(body: CreatePrescriptionBody, user: dict = Depends(require_auth)):
    if user["role"] != "doctor":
        raise HTTPException(status_code=403, detail="Only doctors can create prescriptions")
    if not body.patientId:
        raise HTTPException(status_code=400, detail="patientId is required")

    presc_id = str(uuid.uuid4())
    rows = query(
        """
        INSERT INTO prescriptions (
            id, patient_id, doctor_id, patient_name, doctor_name, doctor_specialization,
            diagnosis, medicines, lab_tests, instructions, follow_up_date, consultation_fee, status
        ) VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,'active')
        RETURNING *
        """,
        (
            presc_id,
            body.patientId,
            str(user["id"]),
            body.patientName,
            user["full_name"],
            user.get("specialization") or "General Physician",
            body.diagnosis,
            json.dumps(body.medicines or []),
            json.dumps(body.labTests or []),
            body.instructions,
            body.followUpDate,
            body.consultationFee,
        ),
    )

    return {"success": True, "prescription": _map_prescription(rows[0])}
